File: data_augmentation/semantic_signatures.r

#setwd("C:/Users/mard019/Desktop/Documents/Conversational AI/RVI Investissement/Voice-bot/")

#source("literals.r", encoding = "utf-8")

Inv_verb_EN <- c("buy"
                 , "contribute"
                 , "convert"
                 , "forward"
                 , "invest"
                 , "make a contribution"
                 , "make a transfer"
                 , "make contribution"
                 , "purchase"
                 , "take"                
                 , "trade"               
                 , "transfer"            
                 , "use"                 
                 , "withdraw"
) 

Inv_account_EN <- c("TFSA"
                    , "Tax-Free Savings Account"
                    , "CELI"
                    , "RRSP"
                    , "RSP"
                    , "RESP")

Reg_account_EN <- c("chequing account"
                    , "checking account"
                    , "savings account"
)

#---# make investment #---#

make_inv_e1 <- function(action_verb = intersect(Inv_verb_EN, c("transfer", "take", "withdraw"))
                        , orig_acc = Reg_account_EN
                        , inv_acc = Inv_account_EN){
  utterances <- do.call(paste, expand.grid(action_verb
                                           , "from my"
                                           , orig_acc
                                           , c("to", "over to", "into", "to put in")
                                           , "my"
                                           , inv_acc))
  return(as.data.frame(utterances))
}


expand.grid(intersect(Inv_verb_EN, c("transfer", "take", "withdraw"))
            , "from my"
            , Reg_account_EN
            , c("to", "over to", "into", "to put in")
            , "my"
            , Inv_account_EN)

make_inv_e2 <- function(action_verb = intersect(Inv_verb_EN, c("transfer", "take", "make a transfer"))
                        , orig_acc = Reg_account_EN
                        , inv_acc = Inv_account_EN
                        , money_exp = setdiff(Money_synonyms_EN, "points")){
  utterances <- do.call(paste, expand.grid(action_verb
                                           , money_exp
                                           , "from my"
                                           , orig_acc
                                           , c("to", "over to", "into", "to put in")
                                           , "my"
                                           , inv_acc))
  return(as.data.frame(utterances))
  
}

make_inv_e3 <- function(action_verb = intersect(Inv_verb_EN, c("transfer", "take"))
                        , inv_acc = Inv_account_EN
                        , money_exp = setdiff(Money_synonyms_EN, "points")){
  utterances <- do.call(paste, expand.grid(action_verb
                                           , money_exp
                                           , c("to", "over to", "into", "to put in")
                                           , "my"
                                           , inv_acc))
  return(as.data.frame(utterances))
}

make_inv_e4 <- function(action_verb = intersect(Inv_verb_EN, c("make a contribution", "contribute", "make contribution", "make a transfer"))
                        , inv_acc = Inv_account_EN){
  utterances <- do.call(paste, expand.grid(action_verb
                                           , "to my"
                                           , inv_acc))
  return(as.data.frame(utterances))
}

make_inv_e5 <- function(action_verb = intersect(Inv_verb_EN, c("use", "trade", "convert"))
                        , inv_acc = Inv_account_EN
                        ,  money_exp = intersect(Money_synonyms_EN, "points")){
  utterances <- do.call(paste, expand.grid(action_verb
                                           , "my"
                                           , money_exp
                                           , "to add to my"
                                           , inv_acc))
  return(as.data.frame(utterances))
}

make_inv_e6 <- function(action_verb = intersect(Inv_verb_EN,c("forward", "trade", "convert"))
                        , inv_acc = Inv_account_EN
                        , money_exp = intersect(Money_synonyms_EN, "points")){
  utterances <- do.call(paste, expand.grid(action_verb
                                           , "my"
                                           , money_exp
                                           , c("towards", "for")
                                           , inv_acc))
  return(as.data.frame(utterances))
}

make_inv_e7 <- function(action_verb = intersect(Inv_verb_EN, "purchase")
                        , inv_acc = c(Inv_account_EN, "investments")){
  utterances <- do.call(paste, expand.grid(c("periodic", "recurring")
                                           , action_verb
                                           , "for my"
                                           , inv_acc))
  return(as.data.frame(utterances))
}

make_inv_e8 <- function(action_verb = intersect(Inv_verb_EN, "buy")
                        , inv_acc = c(Inv_account_EN, "investments")){
  utterances <- do.call(paste, expand.grid(action_verb
                                           , c("funds", "a GIC")
                                           , "from my"
                                           , inv_acc))
  return(as.data.frame(utterances))
}

make_inv_e9 <- function(action_verb = intersect(Inv_verb_EN, "buy")
                        , inv_acc = c(Inv_account_EN, "investments")){
  utterances <- do.call(paste, expand.grid(action_verb
                                           , c("funds", "a GIC")
                                           , "from my"
                                           , inv_acc))
  return(as.data.frame(utterances))
}

make_inv_e10 <- function(action_verb = intersect(Inv_verb_EN, "invest")
                         , inv_acc = Inv_account_EN){
  utterances <- do.call(paste, expand.grid(action_verb
                                           , "in my"
                                           , inv_acc))
  return(as.data.frame(utterances))
}


make_inv_utterances_e1 <- make_inv_e1()
make_inv_utterances_e2 <- make_inv_e2()
make_inv_utterances_e3 <- make_inv_e3()
make_inv_utterances_e4 <- make_inv_e4()
make_inv_utterances_e5 <- make_inv_e5()
make_inv_utterances_e6 <- make_inv_e6()
make_inv_utterances_e7 <- make_inv_e7()
make_inv_utterances_e8 <- make_inv_e8()
make_inv_utterances_e9 <- make_inv_e9()
make_inv_utterances_e10 <- make_inv_e10()

File: semantic_signatures.py
# MAKE_INVESTMENT
# SS1
# TO_ACCOUNT (isA) ACCOUNT_TYPE


# structure 1
SS1_S1 = {  # SS1_S1 = semantic signature 1, structure 1
    "action_verb": [
        "contribute",
        "make a contribution",
        "make a transfer",
        "make contribution",
    ],
    "destination_prep": ["to my"],
    "TO_ACCOUNT": ["TFSA", "Tax-Free Savings Account", "CELI", "RRSP", "RSP", "RESP"],
}
# structure 2
SS1_S2 = {
    "action_verb": ["periodic purchase", "recurring purchase"],
    "destination_prep": ["for my"],
    "TO_ACCOUNT": ["TFSA", "Tax-Free Savings Account", "CELI", "RRSP", "RSP", "RESP"],
}
# structure 3
SS1_S3 = {
    "action_verb": ["buy"],
    "destination_prep": ["funds from my", "a GIC from my"],
    "TO_ACCOUNT": ["TFSA", "Tax-Free Savings Account", "CELI", "RRSP", "RSP", "RESP"],
}
# structure 4
SS1_S4 = {
    "action_verb": ["invest"],
    "destination_prep": ["in my"],
    "TO_ACCOUNT": ["TFSA", "Tax-Free Savings Account", "CELI", "RRSP", "RSP", "RESP"],
}
# structure 5
SS1_S5 = {
    "action_verb": ["transfer", "take"],
    "money_expr": ["money", "some money", "funds", "the funds"],
    "destination_prep": ["to my", "over to my", "into my", "to put in my"],
    "TO_ACCOUNT": ["TFSA", "Tax-Free Savings Account", "CELI", "RRSP", "RSP", "RESP"],
}
# structure 6
SS1_S6 = {
    "action_verb": ["convert", "trade", "use"],
    "money_expr": ["my points"],
    "destination_prep": ["to add to my"],
    "TO_ACCOUNT": ["TFSA", "Tax-Free Savings Account", "CELI", "RRSP", "RSP", "RESP"],
}

# structure 7
SS1_S7 = {
    "action_verb": ["forward", "trade", "convert"],
    "money_expr": ["my points"],
    "destination_prep": ["towards", "for"],
    "TO_ACCOUNT": ["TFSA", "Tax-Free Savings Account", "CELI", "RRSP", "RSP", "RESP"],
}

# MAKE_INVESTMENT
# SS2
# TO_ACCOUNT (isA) ACCOUNT_TYPE
# AMOUNT (isA) nuance_AMOUNT

SS2_S1 = {
    "action_verb": ["transfer", "take"],
    "AMOUNT": ["500 dollars", "500 bucks"],
    "destination_prep": ["to my", "over to my", "into my", "to put in my"],
    "TO_ACCOUNT": ["TFSA", "Tax-Free Savings Account", "CELI", "RRSP", "RSP", "RESP"],
}

# MAKE_INVESTMENT
# SS3
# TO_ACCOUNT (isA) ACCOUNT_TYPE
# FROM_ACCOUNT (isA) ACCOUNT_TYPE

SS3_S1 = {
    "action_verb": ["take", "transfer", "withdraw"],
    "origin_prep": ["from my"],
    "FROM_ACCOUNT": ["chequing account", "checking account", "savings account"],
    "destination_prep": ["to my", "over to my", "into my", "to put in my"],
    "TO_ACCOUNT": ["TFSA", "Tax-Free Savings Account", "CELI", "RRSP", "RSP", "RESP"],
}

SS3_S2 = {
    "action_verb": ["take", "transfer", "withdraw"],
    "money_expr": ["money", "some money", "funds", "the funds"],
    "origin_prep": ["from my"],
    "FROM_ACCOUNT": ["chequing account", "checking account", "savings account"],
    "destination_prep": ["to my", "over to my", "into my", "to put in my"],
    "TO_ACCOUNT": ["TFSA", "Tax-Free Savings Account", "CELI", "RRSP", "RSP", "RESP"],
}

# MAKE_INVESTMENT
# SS4
# TO_ACCOUNT (isA) ACCOUNT_TYPE
# FROM_ACCOUNT (isA) ACCOUNT_TYPE
# AMOUNT (isA) nuance_AMOUNT

SS4_S1 = {
    "action_verb": ["take", "transfer", "withdraw"],
    "AMOUNT": ["500 dollars", "500 bucks"],
    "origin_prep": ["from my"],
    "FROM_ACCOUNT": ["chequing account", "checking account", "savings account"],
    "destination_prep": ["to my", "over to my", "into my", "to put in my"],
    "TO_ACCOUNT": ["TFSA", "Tax-Free Savings Account", "CELI", "RRSP", "RSP", "RESP"],
}
